website/blog/2022-10-31-simple-1.mdx

```diff
@@ -0,0 +1,102 @@
+---
+slug: simple-functions-1
+title: "Simple Functions Introduction and Basic Optimizations"
+authors: [lsakka]
+tags: [tech-blog, simple-functions]
+---
+
+_This blogpost is part of a series of blog posts that discuss different features and optimizations of the simple function interface_.
+
+
+## Introduction to simple functions
+
+Scalar functions are one of the most used extension points in Velox. Since Velox is a vectorized engine, by nature functions are <a href="https://facebookincubator.github.io/velox/develop/scalar-functions.html#vector-functions">"vector functions"</a> that consume vectors and produce vectors. Velox allows users to write functions as vector functions or as single-row operations "simple functions" that are converted to vector functions using template expansion through <a href="https://github.com/facebookincubator/velox/blob/main/velox/expression/SimpleFunctionAdapter.h">SimpleFunctionAdapter</a>.
+
+<figure>
+<img src="/img/simple1_1.png"/>
+</figure>
+
+The user only needs to implement the PlusFunction struct shown in the graph above, which is then expanded using template expansion to a vector function.
+
+Writing simple functions is quite simple and easy; however it's not as powerful as writing vector functions.
```

Comment:
Can we change the order of these ideas? My suggestion would be to start with the paragraph below, talking about some of the issues with vectorized functions (complexity, repetition, reliability), then some of the nice features that simple functions provide, then end with some of the limitations and cases where it's not efficient to use simple functions.

Reply:
I had it like that originally and then changed based on @mbasmanova  comment :

"Would it be possible to rearrange the text to not start with a negative. Maybe first say that writing simple functions is quick and easy. Then, talk about some limitations, then say that vector functions don't have these limitations, but implementing these requires quite a bit more effort."